src/solvers/dg_p4est/containers_2d.jl

```diff
@@ -211,21 +205,119 @@ function init_boundaries_iter_face(info, user_data)
-  # Let p4est iterate over all interfaces and extract boundary data to boundary container
+  # Let p4est iterate over all interfaces and call init_boundaries_iter_face
   iter_face_c = @cfunction(init_boundaries_iter_face, Cvoid, (Ptr{p4est_iter_face_info_t}, Ptr{Cvoid}))
   user_data = [boundaries, 1, mesh]
 
-  GC.@preserve user_data begin
-    p4est_iterate(mesh.p4est,
-                  C_NULL, # ghost layer
-                  # user data [interfaces, boundary_id, mesh]
-                  pointer(user_data),
-                  C_NULL, # iter_volume
-                  iter_face_c, # iter_face
-                  C_NULL) # iter_corner
-  end
+  iterate_faces(mesh, iter_face_c, user_data)
 
   return boundaries
 end
+
+
+# Iterate over all interfaces and extract mortar data to mortar container
+# This function will be passed to p4est in init_mortars! below
+function init_mortars_iter_face(info, user_data)
+  if info.sides.elem_count != 2
+    # Not an inner interface
+    return nothing
+  end
+
+  sides = convert_sc_array(p4est_iter_face_side_t, info.sides)
+
+  if sides[1].is_hanging == false && sides[2].is_hanging == false
+    # Normal interface, no mortar
+    return nothing
+  end
+
+  # Unpack user_data = [mortars, mortar_id, mesh] and increment mortar_id
+  ptr = Ptr{Any}(user_data)
+  data_array = unsafe_wrap(Array, ptr, 3)
+  mortars = data_array[1]
+  mortar_id = data_array[2]
+  data_array[2] += 1
+  mesh = data_array[3]
+
+  # Global trees array
+  trees = convert_sc_array(p4est_tree_t, mesh.p4est.trees)
+  # Quadrant numbering offsets of the quadrants at this interface, one-based indexing
+  offsets = [trees[sides[1].treeid + 1].quadrants_offset,
+             trees[sides[2].treeid + 1].quadrants_offset]
+
+  if sides[1].is_hanging == true
+    # Left is small, right is large
+    small_large = [1, 2]
+
+    local_small_quad_ids = sides[1].is.hanging.quadid
+    # Global IDs of the two small quads
+    small_quad_ids = offsets[1] .+ local_small_quad_ids
+
+    # Just be sure before accessing is.full
+    @assert sides[2].is_hanging == false
+    large_quad_id = offsets[2] + sides[2].is.full.quadid
+  else # sides[2].is_hanging == true
+    # Left is large, right is small
+    small_large = [2, 1]
+
+    local_small_quad_ids = sides[2].is.hanging.quadid
+    # Global IDs of the two small quads
+    small_quad_ids = offsets[2] .+ local_small_quad_ids
+
+    # Just be sure before accessing is.full
+    @assert sides[1].is_hanging == false
+    large_quad_id = offsets[1] + sides[1].is.full.quadid
+  end
+
+  # Write data to mortar container, 1 and 2 are the small elements
+  # p4est uses zero-based indexing; convert to one-based indexing
+  mortars.element_ids[1, mortar_id] = small_quad_ids[1] + 1
+  mortars.element_ids[2, mortar_id] = small_quad_ids[2] + 1
+  # 3 is the large element
+  mortars.element_ids[3, mortar_id] = large_quad_id + 1
+
+  # Face at which the interface lies
+  faces = [sides[1].face, sides[2].face]
```

Comment:
You could consider avoiding these temporary, very small allocations by using tuples everywhere where you only need the data temporarily and only use it within Julia (you cant pass a pointer to a tuple to C). Also above (`small_large`) and in other `*_iter_face` methods that get called a lot (i.e., once per element/face)

Reply:
I am not sure if you see the allocations inside Julia functions in the GC statistics if used as a `@cfunction`, but if you do, you can check the timer outputs whether these suggestions actually make a difference.

Reply:
I haven't optimized all the new mortar code yet. There are more allocations than necessary in multiple places. I wanted to do this once AMR is working.

Reply:
OK, sounds good to me. Maybe add an item to the global TODO issue and then feel free to resolve this conversation

Reply:
Yeah, that sounds good. Some additional performance improvements will probably be possible once I find the time to finish #582